contracts/FeralfileEnglishAuction.sol

```diff
@@ -0,0 +1,378 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./IFeralfileSaleData.sol";
+import "./ECDSASigner.sol";
+import "./IFeralfileVault.sol";
+
+contract IFeralfileExhibitionV4 is IFeralfileSaleData {
+    function buyArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleData calldata saleData_
+    ) external payable {}
+}
+
+contract FeralfileEnglishAuction is Ownable, IFeralfileSaleData, ECDSASigner {
+    struct EnglishAuction {
+        bytes32 id;
+        uint256 startAt;
+        uint256 endAt;
+        uint256 extendDuration;
+        uint256 extendThreshold;
+        uint256 minIncreaseFactor;
+        uint256 minIncreaseAmount;
+        uint256 minPrice;
+        bool isSettled;
+    }
+
+    struct Bid {
+        address bidder;
+        uint256 amount;
+        bool fromFeralFile;
+    }
+
+    struct LatestAuctionStatus {
+        Bid latestBid;
+        uint256 endAt;
+        bool isSettled;
+    }
+
+    mapping(bytes32 => EnglishAuction) public auctions;
+    mapping(bytes32 => Bid) public auctionLatestBid;
+
+    constructor(address signer_) ECDSASigner(signer_) {}
+
+    function getAuctionID(
+        address tokenContractAddr_,
+        uint256[] memory tokenIDs_
+    ) public view returns (bytes32) {
+        // order the tokenIDs based on certain rule
+        for (uint i = 0; i < tokenIDs_.length - 1; i++) {
+            uint minIndex = i;
+            for (uint j = i + 1; j < tokenIDs_.length; j++) {
+                if (tokenIDs_[j] < tokenIDs_[minIndex]) {
+                    minIndex = j;
+                }
+            }
+            (tokenIDs_[i], tokenIDs_[minIndex]) = (
+                tokenIDs_[minIndex],
+                tokenIDs_[i]
+            );
+        }
+
+        // hash the (contractAddr + ordered tokenIDs)
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, tokenContractAddr_, tokenIDs_)
+        );
+
+        // return hash as identifier
+        return message;
+    }
+
+    function auctionOngoing(bytes32 aucId_) public view returns (bool) {
```

Comment:
I'd suggest changing this function name to `ongoingAuction`, the variable could be `id_` and it should be `external`.